### PR TITLE
api, ui: refactor update user

### PR DESCRIPTION
**Pull request by noreng-jg**

### CODE REVIEW

File: api/user/service.go [hunk 1]
```diff
@@ -5,60 +5,88 @@ import (
...
-func (s *service) UpdateDataUser(ctx context.Context, name, username, email, currentPassword, newPassword, ID string) ([]InvalidField, error) {
+func (s *service) UpdateDataUser(ctx context.Context, name, username, email, ID string) ([]InvalidField, error) {
 	var invalidFields []InvalidField
 
-	user, err := s.store.UserGetByID(ctx, ID)
+	_, err := s.store.UserGetByID(ctx, ID)
 
 	if err != nil {
 		return invalidFields, err
 	}
```
Comment (otavio) [inline api/user/service.go:47]:
```suggestion
	if _, err := s.store.UserGetByID(ctx, ID); err != nil {
		return invalidFields, err
	}
```

File: api/user/service.go [hunk 2]
```diff
@@ -5,60 +5,88 @@ import (
...
+	}); err != nil {
+		return invalidFields, ErrBadRequest
 	}
 
-	var checkName, checkEmail bool
+	var checkUsername, checkEmail bool
 
-	user, err = s.store.UserGetByUsername(ctx, username)
+	user, err := s.store.UserGetByUsername(ctx, username)
 	if err == nil && user.ID != ID {
```
Comment (otavio) [inline api/user/service.go:62]:
```go
	if user, err := s.store.UserGetByUsername(ctx, username); err == nil && user.ID != ID {
```

File: api/user/service.go [hunk 3]
```diff
@@ -5,60 +5,88 @@ import (
...
-		checkName = true
-		invalidFields = append(invalidFields, InvalidField{"username", conflictName, "conflict"})
+		checkUsername = true
+		invalidFields = append(invalidFields, InvalidField{"username", ConflictUsername, "conflict"})
 	}
+
 	user, err = s.store.UserGetByEmail(ctx, email)
 	if err == nil && user.ID != ID {
```
Comment (otavio) [inline api/user/service.go:68]:
```suggestion
	if user, err = s.store.UserGetByEmail(ctx, email); err == nil && user.ID != ID {
```

File: api/user/service.go [hunk 4]
```diff
@@ -5,60 +5,88 @@ import (
...
 	}
-	return invalidFields, s.store.UserUpdate(ctx, name, username, email, currentPassword, newPassword, ID)
+
+	return invalidFields, s.store.UserDataUpdate(ctx, name, username, email, ID)
+}
+
+func (s *service) UpdatePasswordUser(ctx context.Context, currentPassword, newPassword, ID string) error {
+	user, err := s.store.UserGetByID(ctx, ID)
+
+	if err != nil {
```
Comment (otavio) [inline api/user/service.go:83]:
```suggestion
	if user, err := s.store.UserGetByID(ctx, ID); err != nil {
```

**Review by otavio — CHANGES_REQUESTED** [review]
There few possible reworks in the code. I didn't check if every use is indeed possible but most should work. Please take a look.

File: api/user/service.go [hunk 5]
```diff
@@ -5,60 +5,88 @@ import (
...
 func NewService(store store.Store) Service {
 	return &service{store}
 }
 
-func (s *service) UpdateDataUser(ctx context.Context, name, username, email, currentPassword, newPassword, ID string) ([]InvalidField, error) {
+func (s *service) UpdateDataUser(ctx context.Context, name, username, email, ID string) ([]InvalidField, error) {
 	var invalidFields []InvalidField
 
-	user, err := s.store.UserGetByID(ctx, ID)
+	_, err := s.store.UserGetByID(ctx, ID)
```
Comment (leonardojoao) [inline api/user/service.go:44]: You can remove this line.

File: api/user/service.go (hunk 4 above)
Comment (noreng-jg) [reply api/user/service.go:83]: In this case user will be in a local scope, but needs to be used in later check

File: api/store/mongo/user_store.go [hunk 6]
```diff
@@ -137,34 +137,32 @@ func (s *Store) UserGetByID(ctx context.Context, ID string) (*models.User, error
 	return user, nil
 }
 
-func (s *Store) UserUpdate(ctx context.Context, name, username, email, currentPassword, newPassword, ID string) error {
-	user, err := s.UserGetByID(ctx, ID)
-	objID, _ := primitive.ObjectIDFromHex(ID)
+func (s *Store) UserDataUpdate(ctx context.Context, data *models.User, ID string) error {
+	objID, err := primitive.ObjectIDFromHex(ID)
```
Comment (leonardojoao) [inline api/store/mongo/user_store.go:142]: You can remove this line.

File: api/store/mongo/user_store.go [hunk 7]
```diff
@@ -137,34 +137,32 @@ func (s *Store) UserGetByID(ctx context.Context, ID string) (*models.User, error
...
+	if _, err := s.UserGetByID(ctx, ID); err != nil {
+		return err
 	}
 
-	if email != "" && email != user.Email {
-		if _, err := s.db.Collection("users").UpdateOne(ctx, bson.M{"_id": objID}, bson.M{"$set": bson.M{"email": email}}); err != nil {
-			return err
-		}
+	objID, err := primitive.ObjectIDFromHex(ID)
+
```
Comment (leonardojoao) [inline api/store/mongo/user_store.go:159]: You can remove this line.

File: api/user/service.go [hunk 8]
```diff
@@ -33,32 +40,66 @@ func NewService(store store.Store) Service {
...
 		return invalidFields, ErrConflict
 	}
-	return invalidFields, s.store.UserUpdate(ctx, name, username, email, currentPassword, newPassword, ID)
+
+	return invalidFields, s.store.UserDataUpdate(ctx, data, ID)
+}
+
+func (s *service) UpdatePasswordUser(ctx context.Context, currentPassword, newPassword, ID string) error {
+	user, err := s.store.UserGetByID(ctx, ID)
+
```
Comment (leonardojoao) [inline api/user/service.go:95]: You can remove this line.

**Pull request merged**